src/consensus/replay_tower.zig

```diff
@@ -4170,24 +4200,104 @@ pub const TestFixture = struct {
+            num_keypairs,
+        );
+
+        for (0..num_keypairs) |_| {
+            try node_pubkeys.append(allocator, Pubkey.initRandom(random));
+            try vote_pubkeys.append(allocator, Pubkey.initRandom(random));
+        }
+
+        self.node_pubkeys = node_pubkeys;
+        self.vote_pubkeys = vote_pubkeys;
+    }
+
     pub fn fill_fork(
         self: *TestFixture,
         allocator: std.mem.Allocator,
         input_tree: Tree,
+        is_frozen: bool,
     ) !void {
+        // Add root to progress map.
+        const gop_root = try self.progress.map.getOrPut(
+            allocator,
+            input_tree.root.slot,
+        );
+        if (gop_root.found_existing) {
+            gop_root.value_ptr.deinit(allocator);
+        }
+        var root_fp = try ForkProgress.zeroes(allocator);
+        root_fp.fork_stats.computed = true;
+        root_fp.fork_stats.my_latest_landed_vote = null;
+        gop_root.value_ptr.* = root_fp;
         // TODO check that root fork exist already and it is being extended
         for (input_tree.data.constSlice()) |tree| {
+            const parent_slot = blk: {
+                const p = tree[1] orelse break :blk 0;
+                break :blk p.slot;
+            };
+            const parent_hash = blk: {
+                const p = tree[1] orelse break :blk Hash.ZEROES;
+                break :blk p.hash;
+            };
+            const element: SlotTracker.Element = .{
+                .constants = .{
+                    .parent_slot = parent_slot,
+                    .parent_hash = parent_hash,
+                    .block_height = tree[0].slot + 1,
+                    .collector_id = Pubkey.ZEROES,
+                    .max_tick_height = 100,
+                    .fee_rate_governor = .DEFAULT,
+                    .epoch_reward_status = .inactive,
+                },
+                .state = .{
+                    .blockhash_queue = .init(.DEFAULT),
+                    .hash = .init(tree[0].hash),
+                    .capitalization = .init(100),
+                    .transaction_count = .init(100),
+                    .signature_count = .init(100),
+                    .tick_height = .init(100),
+                    .collected_rent = .init(100),
+                    .accounts_lt_hash = .init(.{ .data = @splat(100) }),
+                },
+            };
+            try self.slot_tracker.put(allocator, tree[0].slot, element);
             // Populate forkchoice
             try self.fork_choice.addNewLeafSlot(tree[0], tree[1]);
             // Populate progress map
-            var fp = try ForkProgress.zeroes(allocator);
-            fp.fork_stats.computed = true;
-            fp.fork_stats.my_latest_landed_vote = null;
-            try self.progress.map.putNoClobber(
+            const gop = try self.progress.map.getOrPut(
                 allocator,
                 tree[0].slot,
-                fp,
             );
+            if (gop.found_existing) {
```

Comment:
Similar suggestion as above to use `fetchPut` here instead (along with an appropriate `ensureUnusedCapacity(1)` to avoid error paths with redundantly instantiated resources).

Reply:
[9489ab2](https://github.com/Syndica/sig/pull/812/commits/9489ab2eef84ff46f7efdc1d0caf175b9d7a39da)